image-rs/src/unpack.rs

```diff
@@ -39,6 +39,57 @@ fn is_attr_available(path: &Path) -> Result<bool> {
     }
 }
 
+const WHITEOUT_PREFIX: &str = ".wh.";
+const WHITEOUT_OPAQUE_DIR: &str = ".wh..wh..opq";
+
+async fn convert_whiteout(
+    path: &Path,
+    uid: u32,
+    gid: u32,
+    mode: Option<u32>,
+    destination: &Path,
+) -> Result<bool> {
+    // Handle whiteout conversion
+    let name = path
+        .file_name()
+        .unwrap_or_default()
+        .to_str()
+        .ok_or(anyhow!("Invalid unicode in whiteout path: {:?}", path))?;
+
+    if name.starts_with(WHITEOUT_PREFIX) {
+        let parent = path
+            .parent()
+            .ok_or(anyhow!("Invalid whiteout parent for path: {:?}", path))?;
+
+        if name == WHITEOUT_OPAQUE_DIR {
+            let destination_parent = destination.join(parent);
+            xattr::set(destination_parent, "trusted.overlay.opaque", b"y")?;
+            return Ok(false);
+        }
+
+        let original_name = name
+            .strip_prefix(WHITEOUT_PREFIX)
+            .ok_or(anyhow!("Failed to strip whiteout prefix for: {}", name))?;
+        let original_path = parent.join(original_name);
+        let path = CString::new(format!(
+            "{}/{}",
+            destination.display(),
+            original_path.display()
+        ))?;
+
+        let ret = unsafe { nix::libc::mknod(path.as_ptr(), nix::libc::S_IFCHR, 0) };
+        if ret != 0 {
+            bail!("mknod: {:?} error: {:?}", path, io::Error::last_os_error());
+        }
+
+        set_perms_ownerships(&path, ChownType::LChown, uid, gid, mode).await?;
```

Comment:
Do we need also to `set_file_times()` here?

Reply:
Good question.  Looking into the golang base implementation, `set_file_times` is not done. 

The purpose of the whiteout file is to hide the original file. Even if we update the file times, it is not going to be visible in the file system.